### PR TITLE
Feature/ruby tar generation coopr 427

### PR DESCRIPTION
**Pull request by duddun**

### CODE REVIEW

File: lib/provisioner/worker/plugins/automators/chef_solo_automator/chef_solo_automator.rb [hunk 1]
```diff
@@ -51,13 +53,44 @@ def generate_chef_primitive_tar(chef_primitive)
...
+      newfile = gzip(tar(chef_primitive_path))
+      File.new("#{chef_primitive_tar}.new", 'w').write newfile.read
+      # to do: add a test for tarball creation (if !File.exist?("#{chef_primitive_tar}.new") => rescue?)
       `mv "#{chef_primitive_tar}.new" "#{chef_primitive_tar}"`
       log.debug "Generation complete: #{chef_primitive_tar}"
     end
     # rubocop:enable GuardClause
   end
 
+  def tar(path)
```
Comment (dereklwood) [inline lib/provisioner/worker/plugins/automators/chef_solo_automator/chef_solo_automator.rb]: can you copy the same comments describing each of these methods from data-uploader?

File: lib/provisioner/worker/plugins/automators/shell_automator/shell_automator.rb [hunk 2]
```diff
@@ -270,6 +303,7 @@ def stop(inputmap)
   def remove(inputmap)
     runshell(inputmap)
   end
+
```
Comment (dereklwood) [inline lib/provisioner/worker/plugins/automators/shell_automator/shell_automator.rb]: remove blank line

File: lib/provisioner/worker/plugins/automators/chef_solo_automator/chef_solo_automator.rb (hunk 1 above)
Comment (dereklwood) [inline lib/provisioner/worker/plugins/automators/chef_solo_automator/chef_solo_automator.rb]: also, you can actually define these functions in lib/provisioner/plugin/utils.rb (same place as ssh_exec).  Then they are available to all plugins

File: lib/provisioner/worker/plugins/automators/chef_solo_automator/chef_solo_automator.rb [hunk 3]
```diff
@@ -51,13 +53,44 @@ def generate_chef_primitive_tar(chef_primitive)
     # rubocop:disable GuardClause
     if !File.exist?(chef_primitive_tar) || ((Time.now - File.stat(chef_primitive_tar).mtime).to_i > 600)
       log.debug "Generating #{chef_primitive_tar} from #{chef_primitive_path}"
-      `tar -chzf "#{chef_primitive_tar}.new" #{chef_primitive}`
+      newfile = gzip(tar(chef_primitive_path))
+      File.new("#{chef_primitive_tar}.new", 'w').write newfile.read
+      # to do: add a test for tarball creation (if !File.exist?("#{chef_primitive_tar}.new") => rescue?)
```
Comment (dereklwood) [inline lib/provisioner/worker/plugins/automators/chef_solo_automator/chef_solo_automator.rb]: yeah, just do `fail "unable to..." unless File.exist?("#{chef_primitive_tar}.new")`

File: lib/provisioner/worker/plugins/automators/chef_solo_automator/chef_solo_automator.rb [hunk 4]
```diff
@@ -51,7 +53,9 @@ def generate_chef_primitive_tar(chef_primitive)
     # rubocop:disable GuardClause
     if !File.exist?(chef_primitive_tar) || ((Time.now - File.stat(chef_primitive_tar).mtime).to_i > 600)
       log.debug "Generating #{chef_primitive_tar} from #{chef_primitive_path}"
-      `tar -chzf "#{chef_primitive_tar}.new" #{chef_primitive}`
+      newfile = gzip(tar(chef_primitive_path))
+      File.new("#{chef_primitive_tar}.new", 'w').write newfile.read
```
Comment (wolf31o2) [inline lib/provisioner/worker/plugins/automators/chef_solo_automator/chef_solo_automator.rb]: Put `newfile.read` inside parentheses, to make it clear that it's the argument to write.

File: lib/provisioner/worker/plugins/automators/chef_solo_automator/chef_solo_automator.rb [hunk 5]
```diff
@@ -51,7 +53,9 @@ def generate_chef_primitive_tar(chef_primitive)
...
+      fail "unable to generate #{chef_primitive_tar}.new" unless File.exist?("#{chef_primitive_tar}.new")
       `mv "#{chef_primitive_tar}.new" "#{chef_primitive_tar}"`
```
Comment (wolf31o2) [inline lib/provisioner/worker/plugins/automators/chef_solo_automator/chef_solo_automator.rb]: Can you do this move in Ruby, too?

``` ruby
FileUtils.mv("#{chef_primitive_tar}.new", chef_primitive_tar)
```

File: lib/provisioner/worker/plugins/automators/shell_automator/shell_automator.rb [hunk 6]
```diff
@@ -61,7 +63,9 @@ def initialize(env, task)
   def generate_tar(file, path)
     return if File.exist?(file) && ((Time.now - File.stat(file).mtime).to_i < 600)
     log.debug "Generating #{file} from #{path}"
-    `tar -chzf "#{file}.new" -C "#{File.dirname(path)}" #{File.basename(path)}`
+    newfile = gzip(tar_with_resourcename(path))
+    File.new("#{file}.new", 'w').write newfile.read
+    fail "unable to generate #{file}.new" unless File.exist?("#{file}.new")
     `mv "#{file}.new" "#{file}"`
```
Comment (wolf31o2) [inline lib/provisioner/worker/plugins/automators/shell_automator/shell_automator.rb]: Same with this `mv` as above

**Pull request merged**